File: Go-redis/internal/script/seckill.go
package script

import (
	"context"
	_ "embed"
	"go-redis/internal/constant"
	"strconv"

	"github.com/redis/go-redis/v9"
)

//go:embed lua/seckill.lua
var seckillLua string
var seckillScript = redis.NewScript(seckillLua)

//go:embed lua/compensate_seckill.lua
var compensateSeckillLua string
var compensateSeckillScript = redis.NewScript(compensateSeckillLua)

func RunSeckillLua(ctx context.Context, rdb *redis.Client, voucherId, userId uint64) (int64, error) {
	stockKey := constant.SeckillStockKey + strconv.FormatUint(voucherId, 10)
	orderKey := constant.SeckillOrderKey + strconv.FormatUint(voucherId, 10)

	return seckillScript.Run(ctx, rdb, []string{stockKey, orderKey}, userId).Int64()
}

func RunCompensateSeckillLua(ctx context.Context, rdb *redis.Client, voucherId, userId uint64) error {
	stockKey := constant.SeckillStockKey + strconv.FormatUint(voucherId, 10)
	orderKey := constant.SeckillOrderKey + strconv.FormatUint(voucherId, 10)

	return compensateSeckillScript.Run(ctx, rdb, []string{stockKey, orderKey}, userId).Err()
}

File: Go-redis/internal/script/lua/seckill.lua
local stockKey = KEYS[1]
local orderKey = KEYS[2]
local userId = ARGV[1]

-- 判断库存是否充足，不足返回 1
local stock = tonumber(redis.call('GET', stockKey))
if not stock or stock <= 0 then
    return 1
end

-- 若用户已下单，返回 2
if redis.call('SISMEMBER', orderKey, userId) == 1 then
    return 2
end

-- 通过校验后，原子扣库存并记录一人一单
redis.call('INCRBY', stockKey, -1)
redis.call('SADD', orderKey, userId)

-- 抢购资格成功，返回 0
return 0
